src/dmap_import/api_copy_job.py

```diff
@@ -0,0 +1,190 @@
+    )
+
+    insert = postgresql.insert(ApiMetadata).values(
+        url=url,
+        last_updated=result_last_updated,
+    )
+    insert = insert.on_conflict_do_update(
+        constraint="api_metadata_url_key",
+        set_={"last_updated": result_last_updated},
+        where=ApiMetadata.url == url,
+    )
+    db_manager.execute(insert)
+
+
+def run_api_copy(url: str, destination_table: Any) -> None:
+    """
+    Perform DELETE and INSERT on DB destination_table using files retrieved from
+    CUBIC API Endpoint URL
+
+    If exception is thrown during processing of a result, no more results are
+    processed. Results will be re-processed on next event loop in order of
+    'last_updated" result field.
+
+    :param url: CUBIC API Endpoint URL for dataset
+    :param destination_table: SQLAlchemy target table object
+    """
+    db_manager = DatabaseManager()
+
+    for result in get_api_results(url, db_manager):
+        api_result_log = ProcessLogger(
+            "load_api_result",
+            api_url=url,
+            db_table=str(destination_table.__table__),
+            result_url=result["url"],
+            dataset_id=result["dataset_id"],
+            last_updated=result["last_updated"],
+            start_date=result["start_date"],
+            end_date=result["end_date"],
+        )
+        api_result_log.log_start()
+        # temporary local file path for API Result file download
+        temp_file_path = f"/tmp/{url.split('/')[-1]}.csv.gz"
+
+        download_from_url(result["url"], temp_file_path)
+
+        # verify that schema of file recieved from API download matches
+        # expected schema of destination_table
+        # will throw if schemas do not match
+        schema_compare(temp_file_path, destination_table)
+
+        # to prepare for copy of API file download into DB,
+        # delete any records with matching dataset_id in destination_table
+        # also delete any records with no dataset_id, as they should be a
+        # related to a previous processing error
+        delete_dataset_id = sa.delete(destination_table).where(
+            sa.or_(
+                destination_table.dataset_id == result["dataset_id"],
+                destination_table.dataset_id.is_(None),
+            )
+        )
+        delete_result: CursorResult = db_manager.execute(delete_dataset_id)
+
+        copy_local_to_db(temp_file_path, destination_table)
+        db_manager.vaccuum_analyze(destination_table)
+
+        # update dataset_id for all records just loaded into DB from
+        # API downloaded file
+        update_dataset_id = (
+            sa.update(destination_table)
+            .where(destination_table.dataset_id.is_(None))
+            .values(dataset_id=result["dataset_id"])
+        )
+        update_result: CursorResult = db_manager.execute(update_dataset_id)
+
+        insert_update_last_updated(url, result, db_manager)
+
+        # clean up temp API file download
+        os.remove(temp_file_path)
```

Comment:
probably doesn't matter since we're just spinning this ecs task up once a day to run, but something between download and remove throws, this temp file sticks around.

Reply:
added into `finally` block